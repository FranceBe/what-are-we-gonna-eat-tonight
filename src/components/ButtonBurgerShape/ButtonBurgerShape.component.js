import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { spaces, colors, fontSizes } from 'styles/variables'

export const BurgerShape = styled.button`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-end;
  height: ${spaces.large};
  width: ${spaces.large};
  border-radius: ${spaces.large} ${spaces.large} ${spaces.medium} ${spaces.medium};
  background:
  linear-gradient(0deg,
  ${colors.primary_default} 15%,
  ${colors.secondary_default} 27%,
  ${colors.tertiary_default} 47%,
  ${colors.secondary_light} 45%,
  ${colors.secondary_default} 60%,
  ${colors.primary_default} 79%);
  cursor: pointer;
  :focus {
    outline: none;
  }
`
export const ButtonText = styled.span`
  background-color: rgba(255, 255, 255, 0.7);
  padding: ${spaces.small};
  margin-bottom: ${spaces.small};
  font-size: ${fontSizes.px_14}
`

export class ButtonBurgerShape extends React.PureComponent {
  static propTypes = {
    children: PropTypes.any.isRequired,
    onClick: PropTypes.func,
  }

  onActionClick = e => {
    if (this.props.onClick) {
      this.props.onClick()
    } else {
      e.preventDefault()
    }
  }

  render() {
    const { ...rest } = this.props
    return (
      <BurgerShape onClick={this.onActionClick} {...rest}>
        <ButtonText>{this.props.children}</ButtonText>
      </BurgerShape>
    )
  }
}

export default ButtonBurgerShape
