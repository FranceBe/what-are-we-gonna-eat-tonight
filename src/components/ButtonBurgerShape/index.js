import ButtonBurgerShapeComponent,
{ ButtonBurgerShape, BurgerShape, ButtonText }
  from 'components/ButtonBurgerShape/ButtonBurgerShape.component'

export { ButtonBurgerShape, BurgerShape, ButtonText }

export default ButtonBurgerShapeComponent
