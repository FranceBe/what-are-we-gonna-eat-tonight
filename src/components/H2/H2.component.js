import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { fontSizes, spaces, colors } from 'styles/variables'

export const H2Default = styled.h2`
  font-size: ${fontSizes.px_20};
  text-align: center;
  text-shadow: ${spaces.xtraSmall} ${spaces.xtraSmall} ${colors.primary_default};
`

export class H2 extends React.PureComponent {
  static propTypes = {
    children: PropTypes.any,
  }

  render() {
    return (
      <H2Default>{this.props.children}</H2Default>
    )
  }
}

export default H2
