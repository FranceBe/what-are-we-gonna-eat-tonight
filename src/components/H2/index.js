import H2Component, { H2, H2Default } from 'components/H2/H2.component'

export { H2, H2Default }

export default H2Component
