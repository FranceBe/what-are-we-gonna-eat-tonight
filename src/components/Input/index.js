import InputComponent, { Input, InputField, Label, InputContainer } from 'components/Input/Input.component'

export { Input, InputField, Label, InputContainer }

export default InputComponent
