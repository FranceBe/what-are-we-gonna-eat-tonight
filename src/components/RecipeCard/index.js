import RecipeCardComponent, { RecipeCard } from 'components/RecipeCard/RecipeCard.component'

export { RecipeCard }

export default RecipeCardComponent
