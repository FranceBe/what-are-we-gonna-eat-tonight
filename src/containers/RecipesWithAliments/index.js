import RecipesWithAlimentsContainer, {
  RecipesWithAliments,
  InputAndButtonContainer,
  TitleAndButtonsContainer,
  NotFoundMessage,
  ButtonsContainer
} from 'containers/RecipesWithAliments/RecipesWithAliments.container'

export {
  RecipesWithAliments,
  InputAndButtonContainer,
  TitleAndButtonsContainer,
  NotFoundMessage,
  ButtonsContainer
}

export default RecipesWithAlimentsContainer
