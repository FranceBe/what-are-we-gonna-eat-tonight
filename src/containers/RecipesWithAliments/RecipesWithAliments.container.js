import React from 'react'
import styled from 'styled-components'
import H2 from 'components/H2'
import Input, { InputContainer } from 'components/Input'
import ButtonBurgerShape from 'components/ButtonBurgerShape'
import RecipeCard from 'components/RecipeCard'
import { getRandomInt, getRandomNItems } from 'services/random.service'
import { RecipeCardsContainer } from 'components/RecipeCardsContainer'
import { spaces, colors } from 'styles/variables'

export const InputAndButtonContainer = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
  ${InputContainer} {
    flex: 1 1 30%;
  }
`
export const TitleAndButtonsContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  flex: 1 1 30%;
`

export const ButtonsContainer = styled.div`
  display: flex;
  flex-direction: row;
  & button:not(:last-of-type) {
    margin-right: 10px;
  }
`

export const NotFoundMessage = styled.div`
  margin: ${spaces.medium};
  color: ${colors.tertiary_default};
`

export class RecipesWithAliments extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      ingredient: '',
      listOfRecipes: [],
      noRecipeFound: false,
    }
  }

  handleChange = event => {
    if (this.state.noRecipeFound) {
      this.setState({ noRecipeFound: false })
    }
    this.setState({ ingredient: event.target.value })
  }

  getRandomRecipes = n => {
    const randomPage = getRandomInt(50)
    const url = this.state.ingredient ?
      `api/puppy/?q=${this.state.ingredient}&p=${randomPage}` :
      `api/puppy/?p=${randomPage}`
    fetch(url)
      .then(res => res.json())
      .then(data => {
        const listOfRecipes = getRandomNItems(data.results, n)
        this.setState({ listOfRecipes, noRecipeFound: data.results.length === 0 })
      })
      .catch(err => {
        throw Error(err)
      })
  }

  render() {
    return (
      <>
        <InputAndButtonContainer>
          <Input
            label={'Une envie de...'}
            type="text"
            value={this.state.ingredient}
            onChange={this.handleChange}
          />
          <TitleAndButtonsContainer>
            <H2>...combien voulez vous de proppositions ?</H2>
            <ButtonsContainer>
              <ButtonBurgerShape onClick={() => this.getRandomRecipes(1)}> 1 </ButtonBurgerShape>
              <ButtonBurgerShape onClick={() => this.getRandomRecipes(5)}> 5 </ButtonBurgerShape>
              <ButtonBurgerShape onClick={() => this.getRandomRecipes(10)}> 10 </ButtonBurgerShape>
            </ButtonsContainer>
          </TitleAndButtonsContainer>
        </InputAndButtonContainer>
        {this.state.listOfRecipes.length > 0 &&
        <RecipeCardsContainer>
          {this.state.listOfRecipes.map((recipe, i) => (
            <RecipeCard
              key={`recipe-${i}`}
              title={recipe.title}
              ingredients={recipe.ingredients}
              href={recipe.href}
              thumbnail={recipe.thumbnail}
            />
          ))}
        </RecipeCardsContainer>
        }
        {this.state.noRecipeFound &&
        <NotFoundMessage>
          Désolé, nous n&apos;avons trouvé aucune recette qui correspond :&apos;(
        </NotFoundMessage>
        }
      </>
    )
  }
}

export default RecipesWithAliments
