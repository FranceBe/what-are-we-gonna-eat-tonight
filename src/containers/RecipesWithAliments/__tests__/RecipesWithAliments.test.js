import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import RecipesWithAliments, {
  InputAndButtonContainer,
  TitleAndButtonsContainer,
  NotFoundMessage,
  ButtonsContainer
} from 'containers/RecipesWithAliments'
import Input from 'components/Input'
import ButtonBurgerShape from 'components/ButtonBurgerShape'
import H2 from 'components/H2'
import RecipeCard from 'components/RecipeCard'
import * as randomService from 'services/random.service'

describe('RecipesWithAliments', () => {
  const listOfRecipes = [
    {
      title: 'Chocolate Cake',
      href: 'http://siteweb.site.web',
      ingredients: 'chocolat, cake, chocolate cake',
      thumbnail: 'https://upload.wikimedia.org/wikipedia/commons/d/df/Chocolate_cake.jpg',
    },
    {
      title: 'Chocolate Cheese Cake',
      href: 'http://cheese.cake',
      ingredients: 'chocolate, cheese, cake',
      thumbnail: '',
    }
  ]
  let getRandomIntSpy
  beforeAll(() => {
    getRandomIntSpy =
      jest.spyOn(randomService, 'getRandomInt').mockImplementation(() => 15)
    fetch.mockResponse(JSON.stringify({ results: [...listOfRecipes] }))
  })
  afterAll(() => {
    jest.resetAllMocks()
  })
  describe('component style', () => {
    describe('InputAndButtonContainer', () => {
      it('should match snapshot', () => {
        const InputAndButtonContainerTree = renderer.create(<InputAndButtonContainer/>).toJSON()
        expect(InputAndButtonContainerTree).toMatchSnapshot()
      })
    })
    describe('TitleAndButtonsContainer', () => {
      it('should match snapshot', () => {
        const TitleAndButtonsContainerTree = renderer.create(<TitleAndButtonsContainer/>).toJSON()
        expect(TitleAndButtonsContainerTree).toMatchSnapshot()
      })
    })
    describe('NotFoundMessage', () => {
      it('should match snapshot', () => {
        const NotFoundMessageTree = renderer.create(<NotFoundMessage/>).toJSON()
        expect(NotFoundMessageTree).toMatchSnapshot()
      })
    })
    describe('ButtonsContainer', () => {
      it('should match snapshot', () => {
        const ButtonsContainerTree = renderer.create(<ButtonsContainer/>).toJSON()
        expect(ButtonsContainerTree).toMatchSnapshot()
      })
    })
  })
  describe('component rendering', () => {
    describe('given api call has not been done yet', () => {
      it('should have default values for states', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        expect(RecipesWithAlimentsWrapper.state().ingredient).toBe('')
        expect(RecipesWithAlimentsWrapper.state().listOfRecipes).toEqual([])
        expect(RecipesWithAlimentsWrapper.state().noRecipeFound).toBe(false)
      })
      it('should render InputAndButtonContainer, Input with props', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        expect(RecipesWithAlimentsWrapper.find(InputAndButtonContainer)).toHaveLength(1)
        expect(RecipesWithAlimentsWrapper.find(Input)).toHaveLength(1)
        expect(RecipesWithAlimentsWrapper.find(Input).props().label).toBe('Une envie de...')
        expect(RecipesWithAlimentsWrapper.find(Input).props().type)
          .toBe('text')
        expect(RecipesWithAlimentsWrapper.find(Input).props().value)
          .toEqual(RecipesWithAlimentsWrapper.state().ingredient)
        expect(RecipesWithAlimentsWrapper.find(Input).props().onChange)
          .toBe(RecipesWithAlimentsWrapper.instance().handleChange)
      })
      it('should render TitleAndButtonsContainer, H2 and 3 ButtonBurgerShape with props', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        expect(RecipesWithAlimentsWrapper.find(TitleAndButtonsContainer)).toHaveLength(1)
        expect(RecipesWithAlimentsWrapper.find(H2)).toHaveLength(1)
        expect(RecipesWithAlimentsWrapper.find(H2).dive().text())
          .toBe('...combien voulez vous de proppositions ?')
        expect(RecipesWithAlimentsWrapper.find(ButtonBurgerShape)).toHaveLength(3)
        expect(RecipesWithAlimentsWrapper.find(ButtonBurgerShape).at(0).dive().text())
          .toBe(' 1 ')
        expect(RecipesWithAlimentsWrapper.find(ButtonBurgerShape).at(1).dive().text())
          .toBe(' 5 ')
        expect(RecipesWithAlimentsWrapper.find(ButtonBurgerShape).at(2).dive().text())
          .toBe(' 10 ')
      })
      it('should call getRandomRecipes when a BurgerButton is clicked', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        const getRandomRecipesSpy = jest.spyOn(RecipesWithAlimentsWrapper.instance(), 'getRandomRecipes')
        RecipesWithAlimentsWrapper.find(ButtonBurgerShape).at(0).props().onClick()
        expect(getRandomRecipesSpy).toHaveBeenCalledTimes(1)
      })
      it('should not render any RecipeCard', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        expect(RecipesWithAlimentsWrapper.find(RecipeCard)).toHaveLength(0)
      })
    })
    describe('given api call has been done and listOfRecipes is filled', () => {
      it('should render as many RecipeCard as item in listOfRecipes', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        RecipesWithAlimentsWrapper.setState({
          listOfRecipes: [
            {
              title: 'Recipe test',
              ingredients: 'Ingredient, another one, something else',
              href: 'http://recipe.recipe',
              thumbnail: 'http://image.image',
            },
            {
              title: 'Recipe test2',
              ingredients: 'Ingredient, another one, something else2',
              href: 'http://recipe.recipe2',
              thumbnail: 'http://image.image2',
            }
          ]
        })
        expect(RecipesWithAlimentsWrapper.find(RecipeCard)).toHaveLength(2)
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(0).props().title)
          .toBe('Recipe test')
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(0).props().ingredients)
          .toBe('Ingredient, another one, something else')
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(0).props().href)
          .toBe('http://recipe.recipe')
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(0).props().thumbnail)
          .toBe('http://image.image')
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(1).props().title)
          .toBe('Recipe test2')
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(1).props().ingredients)
          .toBe('Ingredient, another one, something else2')
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(1).props().href)
          .toBe('http://recipe.recipe2')
        expect(RecipesWithAlimentsWrapper.find(RecipeCard).at(1).props().thumbnail)
          .toBe('http://image.image2')
      })
    })
    describe('given api respond with nothing', () => {
      it('should render NotFoundMessage', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        RecipesWithAlimentsWrapper.setState({ noRecipeFound: true })
        expect(RecipesWithAlimentsWrapper.find(NotFoundMessage)).toHaveLength(1)
        expect(RecipesWithAlimentsWrapper.find(NotFoundMessage).text())
          .toBe('Désolé, nous n\'avons trouvé aucune recette qui correspond :\'(')
      })
    })
  })
  describe('methods', () => {
    describe('handleChange', () => {
      it('should set noRecipeFound state to false if it was true', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        RecipesWithAlimentsWrapper.setState({ noRecipeFound: true })
        RecipesWithAlimentsWrapper.instance().handleChange({ target: { value: '' } })
        expect(RecipesWithAlimentsWrapper.state().noRecipeFound).toBe(false)
      })
      it('should set ingredient state with the value of event.target.value', () => {
        const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
        RecipesWithAlimentsWrapper.instance().handleChange({ target: { value: 'My Ingredient' } })
        expect(RecipesWithAlimentsWrapper.state().ingredient).toBe('My Ingredient')
      })
    })
    describe('getRandomRecipes', () => {
      describe('ingredient is correctly filled', () => {
        it('should call fetch with api filled with random page and ingredient', () => {
          const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
          RecipesWithAlimentsWrapper.setState({ ingredient: 'chocolate' })
          RecipesWithAlimentsWrapper.instance().getRandomRecipes(10)
          expect(getRandomIntSpy).toHaveBeenCalledTimes(1)
          expect(getRandomIntSpy).toHaveReturnedWith(15) // output is mocked in beforeAll
          expect(fetch).toHaveBeenCalledTimes(1)
          expect(fetch).toHaveBeenCalledWith('api/puppy/?q=chocolate&p=15')
        })
      })
      describe('ingredient is not filled', () => {
        it('should call fetch with api filled with only random page', () => {
          const RecipesWithAlimentsWrapper = shallow(<RecipesWithAliments/>)
          RecipesWithAlimentsWrapper.instance().getRandomRecipes()
          expect(getRandomIntSpy).toHaveBeenCalledTimes(1)
          expect(getRandomIntSpy).toHaveReturnedWith(15) // output is mocked in beforeAll
          expect(fetch).toHaveBeenCalledTimes(1)
          expect(fetch).toHaveBeenCalledWith('api/puppy/?p=15')
        })
      })
    })
  })
})
