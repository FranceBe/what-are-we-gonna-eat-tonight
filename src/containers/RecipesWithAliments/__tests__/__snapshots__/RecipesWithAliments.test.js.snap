// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipesWithAliments component style ButtonsContainer should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 button:not(:last-of-type) {
  margin-right: 10px;
}

<div
  className="c0"
/>
`;

exports[`RecipesWithAliments component style InputAndButtonContainer should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100%;
}

.c0 .c1 {
  -webkit-flex: 1 1 30%;
  -ms-flex: 1 1 30%;
  flex: 1 1 30%;
}

<div
  className="c0"
/>
`;

exports[`RecipesWithAliments component style NotFoundMessage should match snapshot 1`] = `
.c0 {
  margin: 10px;
  color: #EE6663;
}

<div
  className="c0"
/>
`;

exports[`RecipesWithAliments component style TitleAndButtonsContainer should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1 1 30%;
  -ms-flex: 1 1 30%;
  flex: 1 1 30%;
}

<div
  className="c0"
/>
`;
